### PR TITLE
Added zstandard compression tooling, added to bytes codec for JSON serialization

### LINKED ISSUE #321: Add compression approach to Numpy arrays when passed through JSON encoder
It may be possible to reduce the size of `ProtocolDAGResult`s and other `GufeTokenizable`s that require the use of `numpy` arrays for storing numerical data if we apply compression against the byte form of the array, followed by base64 encoding the compressed form for inclusion in the JSON-serialized form. It's unclear *a priori* by how much this will reduce sizes, but may be better than including the data in uncompressed byte form as we do currently.

For high decompression speed, use of [python-zstandard](https://python-zstandard.readthedocs.io/en/latest/concepts.html) might be the best approach.

### PR DESCRIPTION
**Pull request by dotsdl**
Closes #321.

Tips
* Comment "pre-commit.ci autofix" to have pre-commit.ci atomically format your PR.
  Since this will create a commit, it is best to make this comment when you are finished with your work.

Checklist
* [x] Added a ``news`` entry

## Developers certificate of origin
- [x] I certify that this contribution is covered by the MIT License [here](https://github.com/OpenFreeEnergy/openfe/blob/main/LICENSE) and the **Developer Certificate of Origin** at <https://developercertificate.org/>.

### CODE REVIEW

File: gufe/custom_codecs.py [hunk 1]
```diff
@@ -62,8 +64,8 @@ def is_openff_quantity_dict(dct):
 
 BYTES_CODEC = JSONCodec(
     cls=bytes,
-    to_dict=lambda obj: {"latin-1": obj.decode("latin-1")},
-    from_dict=lambda dct: dct["latin-1"].encode("latin-1"),
+    to_dict=lambda obj: {"latin-1": zst_compress(obj).decode("latin-1")},
```
Comment (IAlibay) [inline gufe/custom_codecs.py:67]: The usual question of "will an old version deserialize" 🙃

Comment (IAlibay) [reply gufe/custom_codecs.py:67]: Oops didn't see the try/except TODO!

**Review by IAlibay — COMMENTED** [review]
Do the NPY_DTYPE and NUMPY codecs also need updating here?

File: gufe/tests/test_custom_json.py [hunk 2]
```diff
@@ -113,15 +113,7 @@ def setup_method(self):
...
-        shapes = [
-            [2, 2],
-            [
-                2,
-            ],
-            [
-                3,
-            ],
-        ]
+        shapes = [[2, 2], [2], [3]]
```
Comment (atravitz) [inline gufe/tests/test_custom_json.py:116]: maybe the auto-formatter will revert this, but it was hurting my brain.

File: gufe/tests/test_custom_json.py [hunk 3]
```diff
@@ -184,6 +176,30 @@ def setup_method(self):
...
+                "latin-1": '(µ/ý \ri\x00\x00a test string'
+            }
+        ]
+
+    def test_legacy_uncompressed(self):
+        legacy_codec = JSONCodec(
+            cls=bytes,
+            to_dict=lambda obj: {"latin-1": obj.decode("latin-1")},
+            from_dict=lambda dct: dct["latin-1"].encode("latin-1")
+            )
```
Comment (atravitz) [inline gufe/tests/test_custom_json.py:197]: I can pull this out into a fixture if we want to test in more places, but I kept it simple for now.

Comment (dotsdl) [reply gufe/tests/test_custom_json.py]: Good call; I went ahead and generalized this!

File: gufe/tests/test_custom_json.py (hunk 2 above)
Comment (dotsdl) [reply gufe/tests/test_custom_json.py:137]: Oh shoot, yeah I hope the autoformatter gives something more reasonable!

**Review by dotsdl — COMMENTED** [review]
Looking great @atravitz! I gave this a full review, and generalized `test_legacy_bytes_uncompressed` to apply to all `CustomJSONCodingTest` subclasses.

Can you add a news entry? After that I think this is good to go!

**Comment by dotsdl** [comment]
> Do the NPY_DTYPE and NUMPY codecs also need updating here?

@IAlibay they do not! These convert to bytes, and then the `BYTES` codec is used to convert these into a string representation. So, this PR compresses any `bytes` objects being serialized, and will automatically apply to `NPY_DTYPE` and `NUMPY` objects when we attempt to serialize them.

File: gufe/tests/test_custom_json.py [hunk 4]
```diff
@@ -93,35 +98,43 @@ def _test_round_trip(self, encoder, decoder):
             assert json_str == json_str_2
 
     def test_round_trip(self):
-        encoder, decoder = custom_json_factory([self.codec])
+        encoder, decoder = custom_json_factory(self.required_codecs)
         self._test_round_trip(encoder, decoder)
 
+    def test_legacy_bytes_uncompressed(self):
+        # NOTE: this can be removed in `gufe` 2.0, but is also somewhat harmless
```
Comment (IAlibay) [inline gufe/tests/test_custom_json.py:105]: Please raise an issue with the 2.0 milestone for this.

File: gufe/tests/test_custom_json.py (hunk 2 above)
Comment (IAlibay) [reply gufe/tests/test_custom_json.py:137]:
```suggestion
        shapes = [[2, 2], [2], [3]]  # fmt: skip
```
That way we don't have to rely on the formatter being kind.

**Review by IAlibay — APPROVED** [review]
Couple of things, otherwise lgtm

**Comment by atravitz** [comment]
> Do the NPY_DTYPE and NUMPY codecs also need updating here?

Nope, those use the bytes codec so they don't need to be updated themselves.

File: gufe/tests/test_custom_json.py (hunk 4 above)
Comment (atravitz) [reply gufe/tests/test_custom_json.py:105]: https://github.com/OpenFreeEnergy/gufe/issues/449 opened.

**Review by atravitz — APPROVED** [review]

**Comment by atravitz** [comment]
pre-commit.ci autofix

**Pull request merged**